core/sail/shacl/src/main/java/org/eclipse/rdf4j/sail/shacl/planNodes/LanguageInFilter.java

```diff
@@ -33,12 +40,31 @@ boolean checkTuple(Tuple t) {
 		}
 
 		Optional<String> language = ((Literal) t.getLine().get(1)).getLanguage();
-		return language.filter(languageIn::contains).isPresent();
+		if (!language.isPresent()) {
+			return false;
+		}
+
+		// early matching
+		boolean languageMatches = language.map(String::toLowerCase).filter(normalizedLanguageIn::contains).isPresent();
+		if (languageMatches) {
+			return true;
+		}
+
+		// test according to BCP47
+		String langTag = language.get();
+
+		return checkBCP47CompliantMatch(languageRanges, langTag);
```

Comment:
I think you can DRY this all this up quite a bit by reusing `Literals.isValidLanguageTag`, or `Literals.normalizeLanguageTag` perhaps.

Reply:
The normalize language tag might help.

Reply:
I've ended up just using lowercase for the initial fast language matching and using the SPARQL langMatch code from the StrictEvaluationFactory.